src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1451,19 +1476,45 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
             if (_returnTypesOpt == null &&
                 TryGetReturnType(out TypeWithAnnotations returnType, out FlowAnalysisAnnotations returnAnnotations))
             {
-                TypeWithState returnState;
-                if (node.RefKind == RefKind.None)
+                if (node.RefKind == RefKind.None &&
+                    returnType.Type.SpecialType == SpecialType.System_Boolean &&
+                    isUnconvertedBooleanExpression(expr))
                 {
-                    returnState = VisitOptionalImplicitConversion(expr, returnType, useLegacyWarnings: false, trackMembers: false, AssignmentKind.Return);
+                    // visit the expression without unsplitting, then check parameters marked with flow analysis attributes
+                    Visit(expr);
+
+                    if (this.IsConditionalState)
+                    {
+                        var parameters = this.MethodParameters;
+                        if (!IsConstantFalse(expr))
+                        {
+                            // don't check WhenTrue state on a 'return false;'
+                            checkConditionalParameterState(node.Syntax, parameters, sense: true);
```

Comment:
>sense [](start = 84, length = 5)

It would be good to have a more descriptive name for this parameter. Maybe `possibleReturnValue`?

Reply:
I don't love it either :-$
The reason I used "sense" is because I've seen it used in other places for such situations (applying some logic either to the true or the false branch).